feat: Filtering by status at Backoffice

martintomas opened the pull request:
Adding possibility to filter Projects, PDs and Investors by Status. I have again taken huge inspiration from Tota project ;) and moved full_text filter inside `search_form_for` formular so searching and filters are part of same form.

Styles are really simplified, for example for dropdown, I have used simple selectbox which should be fine from tech point of view and can be replaced later on by something more fancy...

![Screenshot 2022-06-07 at 12 31 55](https://user-images.githubusercontent.com/20660167/172364218-bd1fb2f6-e411-4468-a4fa-ea7aff1effcd.png)

## Testing instructions

Open Backoffice and try to filter by Status Projects, PDs and Investors one by one.

## Tracking

https://vizzuality.atlassian.net/browse/LET-505
https://vizzuality.atlassian.net/browse/LET-531
https://vizzuality.atlassian.net/browse/LET-532
https://vizzuality.atlassian.net/browse/LET-564

martintomas commented on backend/app/services/api/filterer.rb:
```diff
@@ -38,7 +38,8 @@ def apply_full_text_filter
         res[relation] = localized_columns_for relation.to_s.classify.constantize
       end
-      self.query = query.where id: query.dynamic_search(localized_columns_for(query.klass), filters[:full_text], associated_columns).pluck(:id)
+      ids = query.klass.dynamic_search(localized_columns_for(query.klass), filters[:full_text], associated_columns).pluck(:id)
```
ransack and pg_search have problem to work together so I have split query to two separate ones ;)

tsubik replied:
Good to know.

agnessa replied:
It might be possible to use subquery instead of plucking ids? Just in case somebody types, like, "e", and that matches the whole table worth of ids :), maybe it could work almost exactly like you have it, but using select instead of pluck:
fts_subquery = query.klass.dynamic_search(localized_columns_for(query.klass), filters[:full_text], associated_columns).select(:id)

self.query = query.where id: fts_subquery

martintomas replied:
Done, thanks @agnessa ;). I have tried to break query to two by pluck, because that sql is huge and quite complex, but using select has its own advantages, so why not to go for it.

martintomas commented:
> I see you added more filters @martintomas. Can you also add reset filter button? Just a link to entity index page, if any filter is applied.

Sure, I don't think it is at design, but don't see problem with adding it ;)

tsubik commented on backend/app/views/backoffice/projects/_filters.html.erb:
```diff
@@ -0,0 +1,15 @@
+  <%= render partial: "backoffice/shared/full_text_filter", locals: { f: f } %>
+  <div class="w-30 inline-block mx-3">
+    <%= f.input :status_eq, as: :select, required: false, label: false, include_blank: I18n.t("backoffice.common.status"),
+                collection: Project.statuses.map { |k, v| [I18n.t("activerecord.attributes.project.statuses.#{k}"), v] } %>
```
I would add enum model for `ProjectStatus`, and keep the translations there. Maybe frontend will need this as well.

tsubik commented on backend/app/views/backoffice/projects/_filters.html.erb:
```diff
@@ -0,0 +1,15 @@
+  </div>
+  <div class="w-30 inline-block">
+    <%= f.input :category_eq, as: :select, required: false, label: false, include_blank: I18n.t("backoffice.common.category"),
+                collection: Category.all.sort_by(&:name).map { |c| [c.to_s, c.slug] } %>
```
Is the `map` even necessary? I think at least for simple form collection it is not for EnumModels.

tsubik reviewed (approved):
:+1:

agnessa commented on backend/app/models/project.rb:
```diff
@@ -28,7 +28,7 @@ class Project < ApplicationRecord
     :relevant_links
 
-  enum status: {draft: 0, published: 1, closed: 2}, _default: :draft
+  enum status: ProjectStatus::TYPES_WITH_CODE, _default: :draft
```
in my local db project status is still a string field, and so the status filter doesn't work - is there a migration missing or did I mess sth up?

agnessa commented:
I have trouble with testing filtering by mosaic, because I messed up my db and need to reimport the locations table, I'll try to figure that out. In the meantime:

* I found that filtering projects by status doesn't work for me, because the columns a text one in my db rather than int and  I'm not sure if this is sth I messed up locally (I do that a lot sadly) or there's a migration missing to convert it?
* It appears to me as an inconsistency with the status filters that while on projects you have the ability to select a status, apply and then deselect it by choosing "Status", the same does not happen for project developers / investors (include_blank missing - you can ofc reset all filters).
* Also not sure about the "Unverified" status in the projects status dropdown, shouldn't it read "Draft"? so that we don't confuse this status with the trusted / verified flag
* I added a small suggestion regarding using subquery instead of pluck, unimportant if it creates additional problems

martintomas commented:
I have tried to address all your issues @agnessa 

> * I have trouble with testing filtering by mosaic, because I messed up my db and need to reimport the locations table, I'll try to figure that out. In the meantime:

It is integer at my db, so I guess problem is with your database.

> * I found that filtering projects by status doesn't work for me, because the columns a text one in my db rather than int and  I'm not sure if this is sth I messed up locally (I do that a lot sadly) or there's a migration missing to convert it?

Status filter now uses `trusted` column for filtering and hopefully should work for you now.

> * It appears to me as an inconsistency with the status filters that while on projects you have the ability to select a status, apply and then deselect it by choosing "Status", the same does not happen for project developers / investors (include_blank missing - you can ofc reset all filters).

Added missing `include_blank` params so it should be consistent now.

> * Also not sure about the "Unverified" status in the projects status dropdown, shouldn't it read "Draft"? so that we don't confuse this status with the trusted / verified flag

Yup, switched to `trusted` column, text is now verified/unverified.

> * I added a small suggestion regarding using subquery instead of pluck, unimportant if it creates additional problems

Good point, I am using select and subquery now.
